### PR TITLE
doc: update environment helpers in dockerTools docs, add fakeNss section

### PR DESCRIPTION
**Pull request by DanielSidhion**
## Description of changes

This PR updates the content of the environment helpers section in `dockerTools`, adds examples of usage, and follows the docs conventions.
As part of this work, a separate section was created for `fakeNss`, since it was also moved outside `dockerTools`. Proper docs for fakeNss along with examples are also included.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: doc/build-helpers/images/dockertools.section.md [hunk 1]
```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
...
-}
+  copyToRoot = [
+    hello
+    dockerTools.binSh
+  ];
 ```
 
-Creating base files like `/etc/passwd` or `/etc/login.defs` is necessary for shadow-utils to manipulate users and groups.
+After building the image and loading it in Docker, we can create a container based on it and enter a shell inside the container.
+This wouldn't be possible without `binSh`.
```
Comment (roberth) [inline doc/build-helpers/images/dockertools.section.md:972]:
```suggestion
This is made possible by `binSh`.
```

It doesn't have a monopoly on this.

File: doc/build-helpers/images/dockertools.section.md [hunk 2]
```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
...
+GNU bash, version 5.2.21(1)-release (x86_64-pc-linux-gnu)
+(rest of output removed for clarity)
+```
+:::
+
+:::{.example #ex-dockerTools-helpers-buildLayeredImage}
+# Using `dockerTools`'s environment helpers with `buildLayeredImage`
+
+This example adds the [`binSh`](#sssec-pkgs-dockerTools-helpers-binSh) helper to a basic Docker image built with [`dockerTools.buildLayeredImage`](#ssec-pkgs-dockerTools-buildLayeredImage).
+Without this helper, it wouldn't be possible to enter a shell inside a container running this image.
```
Comment (roberth) [inline doc/build-helpers/images/dockertools.section.md:991]:
```suggestion
This helper makes it possible to enter a shell inside the container.
```

As before, and if you want to reinforce the container/image relation, do it earlier, but I don't think it's needed.

File: doc/build-helpers/images/dockertools.section.md [hunk 3]
```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
...
+{ dockerTools, hello }:
+dockerTools.buildLayeredImage {
+  name = "env-helpers";
+  tag = "latest";
 
-  fakeRootCommands = ''
-    ${pkgs.dockerTools.shadowSetup}
-  '';
-  enableFakechroot = true;
+  contents = [
```
Comment (roberth) [inline doc/build-helpers/images/dockertools.section.md:1000]: Not sure that we should recommend this. `contents` has too much implicit behavior for my taste at least.
By making the choice of merging logic explicit, we inform the reader (and our readers' code's readers) of the implementation choice, which they may well have to change.
`symlinkJoin` and `buildEnv` are quite different.

File: doc/build-helpers/images/dockertools.section.md [hunk 4]
```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
...
+
+  config = {
+    Cmd = [ "/bin/hello" ];
+  };
 }
 ```
 
-## fakeNss {#ssec-pkgs-dockerTools-fakeNss}
+After building the image and loading it in Docker, we can create a container based on it and enter a shell inside the container.
+This wouldn't be possible without `binSh`.
```
Comment (roberth) [inline doc/build-helpers/images/dockertools.section.md:1012]:
```suggestion
This is made possible by `binSh`.
```

File: doc/build-helpers/images/dockertools.section.md [hunk 5]
```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
...
+:::
+
+:::{.example #ex-dockerTools-shadowSetup-buildImage}
+# Using `dockerTools.shadowSetup` with `dockerTools.buildImage`
 
-It provides a `/etc/passwd` and `/etc/group`, containing `root` and `nobody`
-users and groups.
+This is an example that shows how to use `shadowSetup` with `dockerTools.buildImage`.
+Note that the extra script in `runAsRoot` uses `groupadd` and `useradd`, which are binaries provided by the `shadow` package.
+These binaries are added to the `PATH` because of the `shadowSetup` script.
```
Comment (roberth) [inline doc/build-helpers/images/dockertools.section.md:1032]:
```suggestion
These binaries are added to the `PATH` by the `shadowSetup` script, but only for the duration of `runAsRoot`.
```

Comment (roberth) [reply doc/build-helpers/images/dockertools.section.md:1032]: This should be added to the earlier text too!

File: doc/build-helpers/images/dockertools.section.md [hunk 6]
```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
...
+:::
+
+:::{.example #ex-dockerTools-shadowSetup-buildLayeredImage}
+# Using `dockerTools.shadowSetup` with `dockerTools.buildLayeredImage`
+
+This is an example that shows how to use `shadowSetup` with `dockerTools.buildLayeredImage`, because it requires `enableFakechroot` to be set to `true`.
+It accomplishes the same thing as [](#ex-dockerTools-shadowSetup-buildImage), but using `buildLayeredImage` instead.
+
+Note that the extra script in `runAsRoot` uses `groupadd` and `useradd`, which are binaries provided by the `shadow` package.
+These binaries are added to the `PATH` because of the `shadowSetup` script.
```
Comment (roberth) [inline doc/build-helpers/images/dockertools.section.md:1065]: Same

File: doc/build-helpers/images/dockertools.section.md [hunk 7]
```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
...
+    WorkingDir = "/data";
+  };
+}
+```
+:::
+
+:::{.example #ex-dockerTools-shadowSetup-buildLayeredImage}
+# Using `dockerTools.shadowSetup` with `dockerTools.buildLayeredImage`
+
+This is an example that shows how to use `shadowSetup` with `dockerTools.buildLayeredImage`, because it requires `enableFakechroot` to be set to `true`.
```
Comment (roberth) [inline doc/build-helpers/images/dockertools.section.md:1061]:
```suggestion
```

Don't document the documentation here. Please move this to the `shadowSetup` docs: When used with `buildLayeredImage`, pass `enableFakechroot`. See example.

File: doc/build-helpers/special.md [hunk 8]
```diff
@@ -3,6 +3,7 @@
 This chapter describes several special build helpers.
 
 ```{=include=} sections
+special/fakenss.section.md
```
Comment (roberth) [inline doc/build-helpers/special.md:6]: I regret that someone at some point named a whole directory so subjectively.
The only thing that's special about Nix stuff is that the outputs are [utterly boring](https://nixos.org/manual/nix/stable/store/file-system-object).

File: doc/build-helpers/special/fakenss.section.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
+# fakeNss {#sec-fakeNss}
+
+Provides `/etc/passwd` and `/etc/group` files that contain `root` and `nobody`, allowing user/group lookups to work in binaries that insist on doing those.
+This might be a better choice than a custom script running `useradd` and related utilities if you only need those files to exist with some entries.
+
+`fakeNss` also provides `/etc/nsswitch.conf`, configuring NSS host resolution to first check `/etc/hosts` before checking DNS, since the default in the absence of a config file (`dns [!UNAVAIL=return] files`) is quite unexpected.
+
+It also creates an empty directory at `/var/empty` because it uses that as the home directory for the `root` and `nobody` users.
```
Comment (roberth) [inline doc/build-helpers/special/fakenss.section.md:8]: I don't think `root` should have its home there, because that makes it all too easy to accidentally put a file there during image creation.

File: doc/build-helpers/special/fakenss.section.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+The `/var/empty` directory can also be used for binaries that use that directory as a `chroot` target to prevent file access in processes that do not need to access files.
```
Comment (roberth) [inline doc/build-helpers/special/fakenss.section.md:9]:
```suggestion
The `/var/empty` directory can also be used as a `chroot` target to prevent file access in processes that do not need to access files, if your container runs such processes.
```

File: doc/build-helpers/images/dockertools.section.md [hunk 11]
```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
...
-}
-```
+Most of these helpers are packages, which means you have to add them to the list of contents to be included in the image (this changes depending on the function you're using to build the image).
+[](#ex-dockerTools-helpers-buildImage) and [](#ex-dockerTools-helpers-buildLayeredImage) show how to include these packages on `dockerTools` functions that build an image.
+For more details on how that works, see the documentation for the function you're using.
 
 ### usrBinEnv {#sssec-pkgs-dockerTools-helpers-usrBinEnv}
 
-This provides the `env` utility at `/usr/bin/env`.
+This provides the `env` utility at `/usr/bin/env` by linking to the `env` binary from the `coreutils` package.
```
[the remark this answers is not in the document]
Comment (DanielSidhion) [reply doc/build-helpers/images/dockertools.section.md:916]: I don't consider the manual as an interface that we commit to, but rather as a digest of the current state of the code (code should be free to change without the constraints of what the docs said before). However, I'm pretty sure some audience might interpret it as something we're committing to, so I'll reword this to prevent that interpretation.

File: doc/build-helpers/images/dockertools.section.md [hunk 12]
```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
...
 
 ### binSh {#sssec-pkgs-dockerTools-helpers-binSh}
 
-This provides `bashInteractive` at `/bin/sh`.
+This provides a `/bin/sh` link to the `bash` binary from the `bashInteractive` package.
```
[the remark this answers is not in the document]
Comment (DanielSidhion) [reply doc/build-helpers/images/dockertools.section.md:920]: Added a note both on docs and in the code to help ensure we don't break this use case.

File: doc/build-helpers/images/dockertools.section.md [hunk 13]
```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
...
+This is a script that sets up files needed for [`shadow`](https://github.com/shadow-maint/shadow) to work (using the `shadow` package from Nixpkgs), and provides all its binaries in `PATH` to allow follow-up scripts to use them.
+It is similar to [`fakeNss`](#sssec-pkgs-dockerTools-helpers-fakeNss), but only sets up a `root` user with different values for the home directory and the shell to use, in addition to setting up files for [PAM](https://en.wikipedia.org/wiki/Linux_PAM) and a {manpage}`login.defs(5)` file.
+Due to how it sets up everything for `shadow` to work, you can add any extra users and/or groups yourself with its utilities.
+
+Since it is a script, this should be used with other dockerTools functions in attributes that expect scripts.
+See [](#ex-dockerTools-shadowSetup-buildImage) and [](#ex-dockerTools-shadowSetup-buildLayeredImage) for how to do that.
+
+:::{.note}
+It is recommended to use either `fakeNss` or `shadowSetup` depending on your use case, but not both.
```
[the remark this answers is not in the document]
Comment (DanielSidhion) [reply doc/build-helpers/images/dockertools.section.md:947]: It depends on how you use them. For example, using both with `buildLayeredImage` doesn't break, but it breaks with `buildImage`. I'll update the language to make it a bit stronger.

File: doc/build-helpers/images/dockertools.section.md (hunk 3 above)
Comment (DanielSidhion) [reply doc/build-helpers/images/dockertools.section.md:1000]: Can you elaborate on this? I'm a bit confused, because there are a couple things going on in my mind:
- `buildImage` uses `copyToRoot` and has `contents` as deprecated, but this example uses `buildLayeredImage` which only provides `contents` for putting stuff into the image. Did you make this comment with this in mind already?
- If you did, I wasn't aware of any scenarios where `contents` in `buildLayeredImage` would cause issues due to implicit behaviour. Can you give me some pointers to learn more about this? This seems like something we definitely want to have documented in the manual, but currently this is not mentioned at all.

File: doc/build-helpers/special.md (hunk 8 above)
Comment (DanielSidhion) [reply doc/build-helpers/special.md:6]: Added this as a note for myself for future work, because it would involve changing a bunch of links too (which is doable, just not something I want to do right now).

File: doc/build-helpers/special/fakenss.section.md (hunk 9 above)
Comment (DanielSidhion) [reply doc/build-helpers/special/fakenss.section.md:8]: This seems like a change to make in the code rather than the docs. For now I could add a note to be careful of this detail, what do you think?

File: doc/build-helpers/images/dockertools.section.md [hunk 14]
```diff
@@ -902,102 +902,192 @@ The `name` argument is the name of the derivation output, which defaults to `fro
...
-Provides `/etc/passwd` and `/etc/group` that contain root and nobody.
-Useful when packaging binaries that insist on using nss to look up
-username/groups (like nginx).
+This is a re-export of the `fakeNss` package from Nixpkgs.
+See [](#sec-fakeNss).
 
 ### shadowSetup {#ssec-pkgs-dockerTools-shadowSetup}
 
-This constant string is a helper for setting up the base files for managing users and groups, only if such files don't exist already. It is suitable for being used in a [`buildImage` `runAsRoot`](#ex-dockerTools-buildImage-runAsRoot) script for cases like in the example below:
+This is a script that sets up files needed for [`shadow`](https://github.com/shadow-maint/shadow) to work (using the `shadow` package from Nixpkgs), and provides all its binaries in `PATH` to allow follow-up scripts to use them.
```
[the remark this answers is not in the document]
Comment (DanielSidhion) [reply doc/build-helpers/images/dockertools.section.md:939]: Restructured the text there to avoid any confusion. Because of that, it might take a bit longer for a reader to understand exactly what the manual is saying, but I linked to the examples which should be the last piece they need to understand things.

File: doc/build-helpers/images/dockertools.section.md (hunk 3 above)
Comment (infinisil) [reply doc/build-helpers/images/dockertools.section.md:1008]: It seems fine to me, let's merge this for now and improve the section with future PRs :)
